### ISSUE TITLE
Allow UO, AuxKernel on linear to be evaluated in Kernel loop

### ISSUE THREAD

**Issue by YaqiWang** [idaholab/moose#9744]
### Description of the enhancement or error report
Some times, those objects can be embedded into the Kernel loop `ComputeResidualThread`. Doing this can potentially reduce one or more element loops per residual evaluation. I propose to add a boolean parameter `in_kernel_loop` for `ElementUserObject`, `SideUserObject`, `InternalSideUserObject`, and `AuxKernel`. When `execute_on` is linear, this parameter can be turned to true to make evaluations of them into kernel loop.

### Rationale for the enhancement or information for reproducing the error
Remove unnecessary assembly overheads. This is an advanced feature similar as the `compute` for `Material`.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
New capability.

**Comment by permcody** [comment]
Yes, we are planning on these kinds of optimizations very soon, but we want to do this right. I don't really want multiple ways of calculating the same quantities so as we consolidate, we'll want to remove loops. This is actually a very tough task but we're hoping it yields some decent gains. First things first though. We need timing information which is what we are working on now in preparation for these optimizations.
